nova/virt/vmwareapi/vm_util.py

```diff
@@ -839,6 +839,66 @@ def get_hardware_devices(session, vm_ref):
     return vim_util.get_array_items(hardware_devices)
 
 
+def _is_before_in_boot_order(disk1, disk2):
+    if not disk2:
+        return False
+    if not disk1:
+        return True
+    return (disk1.controllerKey > disk2.controllerKey
+        or disk1.controllerKey == disk2.controllerKey
+        and disk1.unitNumber > disk2.unitNumber)
+
+
+def get_hardware_devices_by_type(session, vm_ref):
+    nics = {}
+    controller = {}
+    disks = {}
+    swap = []
+    ephemeral = []
+    root_device = None
+
+    for device in get_hardware_devices(session, vm_ref):
+        device_type = device.__class__.__name__
+        if device_type in ALL_SUPPORTED_NETWORK_DEVICES:
+            mac_address = getattr(device, 'macAddress', None)
+            if not mac_address:
+                LOG.warning("Could not ge mac address of NIC {}", device.key)
+                continue
+            if mac_address in nics:
+                LOG.warning("Multiple NICs with the same MAC")
+            else:
+                nics[mac_address] = device
+        elif device_type in CONTROLLER_TO_ADAPTER_TYPE:
+            controller[int(device.key)] = device
+        elif device_type == "VirtualDisk":
+            backing_type = device.backing.__class__.__name__
+            if backing_type == "VirtualDiskFlatVer2BackingInfo":
+                if "swap" in device.backing.fileName:
+                    swap.append(device)
+                elif "ephemeral" in device.backing.fileName:
+                    ephemeral.append(device)
+                else:
+                    if _is_before_in_boot_order(root_device, device):
+                        root_device = device
```

Comment:
its not clear from name is_first_before_second_in_boot_order or is_second_before_first_in_boot_order ??

Reply:
True, the order is non-standard...
I would suggest to follow the `_in_boot_order(a, b) -> if a is before b in boot order`

Reply:
yes, makes sense